feat: support `no_std` for `reth-storage-errors`

JackG-eth opened the pull request:
Description

have a subset of crates with no_std supports (evm specifically)
we need to do some prep for this first by making some crates no_std first

Motivation
closes #8459

mattsse commented on crates/storage/errors/src/db.rs:
```diff
@@ -1,57 +1,97 @@
+use alloc::boxed::Box;
+
+#[cfg(feature = "std")]
+use std::{error::Error, fmt};
```
I prefer using the fully qualified path where necessary and not feature gate the `use` import

JackG-eth replied:
Is this from a readability perspestive?

mattsse replied:
yes and reduces cfgs

mattsse commented on crates/storage/errors/src/db.rs:
```diff
@@ -1,57 +1,97 @@
+use alloc::boxed::Box;
+
+#[cfg(feature = "std")]
+use std::{error::Error, fmt};
```
fmt should be under core::fmt

mattsse commented on crates/storage/errors/src/db.rs:
```diff
@@ -1,57 +1,97 @@
+impl Display for DatabaseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result {
+        match self {
+            Self::Open(info) => f.write_fmt(format_args!("failed to open the database: {0}", info)),
```
I believe this can be replace with
`write!(f, "", args)`

mattsse commented on crates/storage/errors/src/provider.rs:
```diff
@@ -154,17 +240,24 @@ pub struct RootMismatch {
     pub block_hash: BlockHash,
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for RootMismatch {}
```
like this

mattsse commented on crates/storage/errors/src/db.rs:
```diff
@@ -1,57 +1,79 @@
+    str::FromStr,
+};
+
+extern crate alloc;
```
this should be in lib.rs

mattsse commented on crates/storage/errors/src/db.rs:
```diff
@@ -169,7 +211,7 @@ impl LogLevel {
     type Err = String;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
```
```suggestion
    fn from_str(s: &str) -> core::result::Result<Self, Self::Err> {
```

mattsse commented on crates/storage/errors/src/provider.rs:
```diff
@@ -176,3 +266,19 @@ impl From<ConsistentViewError> for ProviderError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Self::Syncing { best_block } => {
+                f.write_fmt(format_args!("node is syncing. best block: {0:?}", best_block,))
+            }
+            Self::Inconsistent { tip } => {
+                f.write_fmt(format_args!("inconsistent database state: {0:?}", tip))
```
write!

mattsse reviewed (changes requested):
last few nits

mattsse reviewed (changes requested):
we don't have proper no-std tests in ci yet, so this currently passes, but fails with `cargo c --no-default-features`, mainly std related imports

JackG-eth commented:
> we don't have proper no-std tests in ci yet, so this currently passes, but fails with `cargo c --no-default-features`, mainly std related imports

added necessary imports where needed, also added a couple more feature std flags where std libs are used.

mattsse reviewed (approved):
@DaniPopes I believe no-std-thiserror has a few limitations.

rolling this back once Error is available in no_std isn't a big deal so I'm fine with this, even if only temporary
wdyt?
